src/Model_ELBDM/CPU_ELBDM/CPU_ELBDMSolver_FFT.cpp

```diff
@@ -0,0 +1,250 @@
+// Function    :  Psi_Advance_FFT
+// Description :  Use FFT to advance the wave function by the kinetic operator (the pseudo-spectral method)
+//
+// Note        :  1. Invoked by CPU_ELBDMSolver_FFT()
+//                2. Advance wave function by exp( -i*dt*k^2/(2*ELBDM_ETA) ) in the k-space
+//
+// Parameter   :  PsiR     : Array storing the real part of wave function (input and output)
+//                PsiI     : Array storing the imag part of wave function (input and output)
+//                j_start  : Starting j index
+//                dj       : Size of array in the j (y) direction after the forward FFT
+//                PsiK_Size: Size of the array "PsiK"
+//                dt       : Time interval to advance solution
+//-------------------------------------------------------------------------------------------------------
+void Psi_Advance_FFT( real *PsiR, real *PsiI, const int j_start, const int dj, const int PsiK_Size, const real dt )
+{
+
+   const int Nx        = NX0_TOT[0];
+   const int Ny        = NX0_TOT[1];
+   const int Nz        = NX0_TOT[2];
+   const real dh       = amr->dh[0];
+   const real Dt_2Eta  = (real)0.5*dt/ELBDM_ETA;
+   real PsiKR, PsiKI, DtKK_2Eta;
+   fftw_complex *PsiK;
+   PsiK = (fftw_complex*) malloc( PsiK_Size * sizeof(fftw_complex));
+
+   for (int t=0; t<PsiK_Size; t++)
+   {
+      PsiK[t].re = PsiR[t];
+      PsiK[t].im = PsiI[t];
+   }
+
+
+// forward FFT
+#  ifdef SERIAL
+   fftwnd_one( FFTW_Plan_Psi, PsiK, NULL );
+#  else
+   fftwnd_mpi( FFTW_Plan_Psi, 1, PsiK, NULL, FFTW_TRANSPOSED_ORDER );
+#  endif
+
+
+// set up the dimensional wave number
+// and the k-space evolution phase (K_i*K_i)*dt/(2*ELBDM_ETA), where i=x,y,z
+   real Kx[Nx], Ky[Ny], Kz[Nz];
+   real DtKxKx_2Eta[Nx], DtKyKy_2Eta[Ny], DtKzKz_2Eta[Nz];
+
+   for (int i=0; i<Nx; i++)
+   {
+      Kx[i]          = ( i <= Nx/2 ) ? 2.0*M_PI/(Nx*dh)*i : 2.0*M_PI/(Nx*dh)*(i-Nx);
+      DtKxKx_2Eta[i] = SQR(Kx[i])*Dt_2Eta;
+   }
+   for (int j=0; j<Ny; j++)
+   {
+      Ky[j]          = ( j <= Ny/2 ) ? 2.0*M_PI/(Ny*dh)*j : 2.0*M_PI/(Ny*dh)*(j-Ny);
+      DtKyKy_2Eta[j] = SQR(Ky[j])*Dt_2Eta;
+   }
+   for (int k=0; k<Nz; k++)
+   {
+      Kz[k]          = ( k <= Nz/2 ) ? 2.0*M_PI/(Nz*dh)*k : 2.0*M_PI/(Nz*dh)*(k-Nz);
+      DtKzKz_2Eta[k] = SQR(Kz[k])*Dt_2Eta;
+   }
+
+
+// multiply the wave function by exp( -i*dt*k^2/(2*ELBDM_ETA) ) in the k-space
+   long ID;
+#  ifdef SERIAL // serial mode
+
+   for (int k=0; k<Nz; k++)
+   {
+      for (int j=0; j<Ny; j++)
+      for (int i=0; i<Nx; i++)
+      {
+         ID = ((long)k*Ny + j)*Nx + i;
+
+#  else // parallel mode
+   int j;
+
+   for (int jj=0; jj<dj; jj++)
+   {
+      j = j_start + jj;
+
+      for (int k=0; k<Nz; k++)
+      for (int i=0; i<Nx; i++)
+      {
+         ID = ((long)jj*Nz + k)*Nx + i;
+
+#  endif // #ifdef SERIAL ... else ...
+
+         PsiKR = PsiK[ID].re;
+         PsiKI = PsiK[ID].im;
+         DtKK_2Eta = DtKxKx_2Eta[i] + DtKyKy_2Eta[j] + DtKzKz_2Eta[k];
```

Comment:
Add `const real` if we want to support OpenMP here.

Reply:
Done.